### ISSUE TITLE
Backward hooks of discriminator parameters are triggered twice when training GAN

### ISSUE THREAD

**Issue by hughplay** [uber/horovod#796]
## Background

I'm using Horovod and PyTorch to train a GAN. The key code is showing below and more detail about background can be found in #784 :

Optimizer definition:
``` py 
self.generator_optimizer = get_optimizer(self.config, self.generator.parameters(), hvd)
self.discriminator_optimizer = get_optimizer(self.config, self.discriminator.parameters(), hvd)
self.generator_optimizer = hvd.DistributedOptimizer(self.generator_optimizer, named_parameters=self.generator.named_parameters())
self.discriminator_optimizer = hvd.DistributedOptimizer(self.discriminator_optimizer, named_parameters=self.discriminator.named_parameters())
```
Training:
``` py
self.generator.train()
self.discriminator.train()

gen_imgs = self.generator(sketches)

self.generator_optimizer.zero_grad()
g_loss = self.criterion(as_real=self.discriminator(gen_imgs))
g_loss.backward()
self.generator_optimizer.step()

self.discriminator_optimizer.zero_grad()
d_loss = self.criterion(as_real=self.discriminator(imgs), as_fake=self.discriminator(gen_imgs.detach()))
d_loss.backward()
self.discriminator_optimizer.step()
```

The problem is that when run the code above, it will raised Assertion Error:
https://github.com/uber/horovod/blob/42c0977d8ee62060b1d4c25a98b15c9cac85ebef/horovod/torch/__init__.py#L92

This hook is registered to parameters when define `hvd.DistributedOptimizer()`, and triggered when calling `loss.backward()`.

## Guessing about the reason of assertion error

When calling `g_loss.backward()`(generator first and then discriminator), the gradients of `discriminator` are also computed and then the hooks bind to those  parameters will be triggered. After that, `d_loss.backward()` will trigger the hooks of discriminator's parameters once again, raise Assertion Error.

## Two ways to make it work

I try two ways to make things work, first I add `self.generator_optimizer.zero_grad()` after generator optimized:
``` py
self.generator_optimizer.zero_grad()
g_loss = self.criterion(as_real=self.discriminator(gen_imgs))
g_loss.backward()
self.generator_optimizer.step()

self.generator_optimizer.zero_grad()
self.discriminator_optimizer.zero_grad()
d_loss = self.criterion(as_real=self.discriminator(imgs), as_fake=self.discriminator(gen_imgs.detach()))
d_loss.backward()
self.discriminator_optimizer.step()
```
This doesn't work.

And I try flip `requires_grad` of discriminator parameters before and after optimizing generator.

``` py
for param in self.discriminator.parameters():
    param.requires_grad = False              
self.generator_optimizer.zero_grad()
g_loss = self.criterion(as_real=self.discriminator(gen_imgs))
g_loss.backward()
self.generator_optimizer.step()
for param in self.discriminator.parameters():
    param.requires_grad = True     

self.generator_optimizer.zero_grad()
self.discriminator_optimizer.zero_grad()
d_loss = self.criterion(as_real=self.discriminator(imgs), as_fake=self.discriminator(gen_imgs.detach()))
d_loss.backward()
self.discriminator_optimizer.step()
```
This seems work fine.

## Question

So the question is whether the second way is correct?
It seems this is the only way to avoid unnecessary `allreduce_async_` among those parameters that not need to optimized.

And the `hook` of version `0.15.2` and `master` seems different, and `hook` in `master` doesn't `assert p not in self._handles` anymore.

I'm sorry for the long question, it's hard to make it clear in several words for me.

**Comment by alsrgv** [comment]
@hughplay, it's correct to flip requires_grad like you're doing in the second way to train GAN.  We have a test case demonstrating similar concept here: https://github.com/uber/horovod/blob/master/test/test_torch.py#L1163

The code you're referring to was changed to support multiple backwards passes per `optimizer.step()` - but that's aimed to support local gradient aggregation rather than training GANs.

**Comment by cheneyoung** [comment]
@hughplay hi，recently，I use horovod to train UGATIT， but there is no effect for multiply GPU to train the model. Is there any effect for you training the GAN?
